[senseBox] Switch caching to ESH ExpiringCacheMap

hakan42 opened the pull request:
* Removed indivial caching class (which was adapted from Yahoo Weather anyway)
* Extracted data fetching to SenseBoxAPIConnection class
* Switched HTTP client to ESH HttpUtil

Signed-off-by: Hakan Tandogan <hakan@tandogan.com>

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -34,36 +29,38 @@
 
     ScheduledFuture<?> refreshJob;
 
     private final int CACHE_EXPIRY = 10 * 1000; // 10s
```
This seems like a real constant, it should be static

hakan42 replied:
Agreed, has been a typo and fixed.

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -34,36 +29,38 @@
 
+    private final ExpiringCacheMap<String, SenseBoxData> CACHE = new ExpiringCacheMap<String, SenseBoxData>(CACHE_EXPIRY);
+
+    private final String CACHE_KEY_DATA = "DATA";
```
This could be a real constant make it static

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -34,36 +29,38 @@
 
     private final int CACHE_EXPIRY = 10 * 1000; // 10s
 
+    private final ExpiringCacheMap<String, SenseBoxData> CACHE = new ExpiringCacheMap<String, SenseBoxData>(CACHE_EXPIRY);
```
This is not a constant, make it lowercase

hakan42 replied:
Agreed, and will fix, but maybe we should change YahooWeather as well to have a consistent handling of the cache overall:

https://github.com/eclipse/smarthome/pull/3628

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/internal/SenseBoxAPIConnection.java:
```diff
@@ -0,0 +1,151 @@
+    public SenseBoxAPIConnection() {
+        String openHABVersion = "2.1.0-SNAPSHOT";
+        Version version = FrameworkUtil.getBundle(this.getClass()).getVersion();
+        // TODO is there a way to obtain the OH version?
```
There is a TODO, imho it should be completed before merging the PR

hakan42 replied:
Yes, there is, I did not find a way to obtain the OH version programmatically. I'm typing a question about this in the forums just right now 😄

martinvw replied:
I just added a suggestion ;-)

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/internal/SenseBoxAPIConnection.java:
```diff
@@ -0,0 +1,151 @@
+
+    private Gson gson = new Gson();
+
+    private final static Properties HEADERS = new Properties();
```
Non-preferred order of modifiers correct order is `private static final` See also: http://stackoverflow.com/questions/16731240/what-is-a-reasonable-order-of-java-modifiers-abstract-final-public-static-e#16732059

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/internal/SenseBoxAPIConnection.java:
```diff
@@ -0,0 +1,151 @@
+                    parsedData.setHumidity(sensor);
+                } else if ("°C".equals(sensor.getUnit())) {
+                    parsedData.setTemperature(sensor);
+                } else {
```
We could make this an `else if (logger.isDebugEnabled())`

hakan42 replied:
agreed, and fixed.

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/internal/SenseBoxAPIConnection.java:
```diff
@@ -0,0 +1,151 @@
+        } catch (IOException e) {
+            logger.debug("IO problems while fetching data: {} / {}", query, e.getMessage());
+            // TODO how would we switch the caller to OFFLINE???
+            // updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
This looks WIP, maybe you should register add a listener to this connection which can report back.

hakan42 replied:
Fixed. The data object hold a status property now which can be evaluated by the handler.

hakan42 commented:
@martinvw , I believe I addressed all your points except the one TODO about the OH / ESH version for which I have a question still open in the forums.

hakan42 commented:
@martinvw , I have fixed the TODO with the OH version as well. From my point of view, this PR is ready for consideration. Could you give it another look please?
